File: 74-Search-a-2D-Matrix/solution.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        len_m = len(matrix)
        len_n = len(matrix[0])
        start, end = 0, len_m * len_n - 1
        while start <= end:
            mid = (start + end) / 2
            value = matrix[mid / len_n][mid % len_n]
            if value == target:
                return True
            elif value > target:
                end = mid - 1
            else:
                start = mid + 1
        return False
        

File: 74-Search-a-2D-Matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty()) {
            return false;
        }
        
        if (matrix[0].empty()) {
            return false;
        }

        const int m = matrix.size();
        const int n = matrix[0].size();
        int left = 0;
        int right = m * n;

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (matrix[mid / n][mid % n] >= target) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        if (left != m * n && matrix[left / n][left % n] == target) {
            return true;
        }

        return false;
    }
};